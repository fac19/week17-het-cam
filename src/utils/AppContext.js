import React, { useState, createContext } from "react";

const AppContext = createContext();

function AppContextProvider(props) {
  const [userInfo, setUserInfo] = useState();
  const [loginInfo, setLoginInfo] = useState({
    name: "",
    email: "",
  });
  const [isUserInfoComplete, setIsUserInfoComplete] = useState(false);
  const [missionState, setMissionState] = useState("get");

  const providerValue = {
    userInfo,
    setUserInfo,
    loginInfo,
    setLoginInfo,
    isUserInfoComplete,
    setIsUserInfoComplete,
    missionState,
    setMissionState,
  };

  return (
    <AppContext.Provider value={providerValue}>
      {props.children}
    </AppContext.Provider>
  );
}

export { AppContext, AppContextProvider };
