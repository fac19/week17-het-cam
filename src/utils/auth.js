import React from "react";
import Landing from "../pages/landingPage/Landing";

export default function requireAuthentication(Component, user) {
  function AuthHOC(props) {
    return user ? <Component user={user} {...props} /> : <Landing />;
  }
  return AuthHOC;
}
