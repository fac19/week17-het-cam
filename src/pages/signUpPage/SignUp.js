import React, { useContext } from "react";
import { useHistory } from "react-router-dom";
import Heading from "../../components/global/heading/Heading";
import {
  DropdownCountries,
  DropdownGender,
} from "../../components/global/forms/dropdown/Dropdown";
import DateInput from "../../components/global/forms/dateInput/DateInput";
import { TextButton } from "../../components/global/buttons/Buttons";
import { FormContainer, FormInputWrapper } from "./SignUp.style";
import { AppContext } from "../../utils/AppContext";

const SignUp = () => {
  const { isUserInfoComplete, setIsUserInfoComplete } = useContext(AppContext);
  const { loginInfo, setLoginInfo } = useContext(AppContext);

  const history = useHistory();
  const token = JSON.parse(localStorage.getItem("token"));

  const name = token ? token.full_name : "";
  const email = token ? token.email : "";

  const [form, setForm] = React.useState({
    Name: name,
    Email: email,
    dob: "",
    gender: "",
    country: "",
  });

  const handleChange = (event) => {
    const { name, value } = event.target;
    setForm({ ...form, [name]: value });
  };

  const handleSubmit = async (event) => {
    event.preventDefault();

    await fetch("../../../.netlify/functions/post-student/post-student.js", {
      method: "POST",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify(form),
    });
    history.push("/my-missions");
  };

  // Retreive data from airtable, then check if user is already registered with all neccessary information
  React.useEffect(() => {
    // get user email from state
    let userEmail = loginInfo.email;

    const func = async () => {
      const post = await fetch(
        `../../../.netlify/functions/fetch-student/fetch-student.js?email=${userEmail}`,
      );

      const finalFetch = await post.json();
      setIsUserInfoComplete(finalFetch);
    };

    func();
  }, [loginInfo, setIsUserInfoComplete]);

  React.useEffect(() => {
    if (isUserInfoComplete) {
      return history.push("/my-missions");
    }
  }, [isUserInfoComplete, history]);

  React.useEffect(() => {
    const token = JSON.parse(localStorage.getItem("token"));
    if (token) {
      setLoginInfo({
        email: token.email,
        name: token.full_name,
      });
    }
  }, [setLoginInfo]);

  return (
    <>
      <Heading>Before you begin...</Heading>
      <FormContainer onSubmit={handleSubmit}>
        <FormInputWrapper>
          <DateInput onChange={handleChange} />
        </FormInputWrapper>
        <FormInputWrapper>
          <DropdownCountries onChange={handleChange} />
        </FormInputWrapper>
        <FormInputWrapper>
          <DropdownGender onChange={handleChange} />
        </FormInputWrapper>
        <TextButton type={"submit"} text={"Get started!"} />
      </FormContainer>
    </>
  );
};

export default SignUp;
