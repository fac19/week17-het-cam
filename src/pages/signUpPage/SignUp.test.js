import React from "react";
import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { AppContextProvider } from "../../utils/AppContext";
import SignUp from "./SignUp";

test("SignUp page renders correctly", () => {
  render(
    <AppContextProvider>
      <SignUp type={"submit"} text={"Get started!"}></SignUp>
    </AppContextProvider>,
  );
  screen.getByText("Get started!");
});

test("SignUp page is styled", () => {
  const tree = renderer
    .create(
      <AppContextProvider>
        <SignUp />
      </AppContextProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
