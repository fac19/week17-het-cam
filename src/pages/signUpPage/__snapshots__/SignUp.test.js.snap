// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp page is styled 1`] = `
Array [
  .c0 {
  font-family: "Josefin Sans",sans-serif;
  font-size: 2rem;
  color: #032049;
  text-align: center;
}

@media (min-width:600px) {
  .c0 {
    font-size: 4rem;
  }
}

@media (min-width:768px) {

}

<h1
    className="c0"
  >
    Before you begin...
  </h1>,
  .c2 {
  text-align: center;
  padding: 0.5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 70vh;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 2rem;
}

@media (min-width:600px) {

}

@media (min-width:768px) {
  .c0 {
    height: 50vh;
  }
}

<form
    className="c0"
    onSubmit={[Function]}
  >
    <div
      className="c1"
    >
      <label
        className="c2"
        htmlFor="dob"
      >
        Date of birth
      </label>
      <input
        id="dob"
        name="dob"
        onChange={[Function]}
        type="date"
      />
    </div>
    <div
      className="c1"
    >
      <label
        className="c2"
        htmlFor="countries"
      >
        Country
      </label>
      <select
        defaultValue="select-option"
        id="country"
        name="country"
        onChange={[Function]}
        required={true}
      >
        <option
          disabled={true}
          value="select-option"
        >
          Select option
        </option>
        <option
          value="Nepal"
        >
          Nepal
        </option>
        <option
          value="India"
        >
          India
        </option>
        <option
          value="Rwanda"
        >
          Rwanda
        </option>
      </select>
    </div>
    <div
      className="c1"
    >
      <label
        className="c2"
        htmlFor="gender"
      >
        Gender
      </label>
      <select
        defaultValue="select-option"
        id="gender"
        name="gender"
        onChange={[Function]}
        required={true}
      >
        <option
          disabled={true}
          value="select-option"
        >
          Select option
        </option>
        <option
          value="Female"
        >
          Female
        </option>
        <option
          value="Male"
        >
          Male
        </option>
        <option
          value="Other"
        >
          Other
        </option>
        <option
          value="Prefer Not To Say"
        >
          Prefer not to say
        </option>
      </select>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-root-2"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
      <span
        className="MuiButton-label"
      >
        Get started!
      </span>
    </button>
  </form>,
]
`;
